File: src/test/java/de/mrapp/apriori/metrics/ConvictionTest.java
/*
 * Copyright 2017 Michael Rapp
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package de.mrapp.apriori.metrics;

import de.mrapp.apriori.AssociationRule;
import de.mrapp.apriori.ItemSet;
import de.mrapp.apriori.NamedItem;
import org.junit.Test;

import static org.junit.Assert.assertEquals;

/**
 * Tests the functionality of the class {@link Conviction}.
 *
 * @author Michael RappO
 */
public class ConvictionTest {

    /**
     * Tests the functionality of the evaluate-method.
     */
    @Test
    public final void testEvaluate() {
        double bodySupport = 0.7;
        ItemSet<NamedItem> body = new ItemSet<>();
        body.add(new NamedItem("a"));
        body.setSupport(bodySupport);
        double headSupport = 0.8;
        ItemSet<NamedItem> head = new ItemSet<>();
        head.add(new NamedItem("b"));
        head.setSupport(headSupport);
        double support = 0.5;
        AssociationRule<NamedItem> rule = new AssociationRule<>(body, head, support);
        double confidence = new Confidence().evaluate(rule);
        assertEquals((1 - headSupport) / (1 - confidence), new Conviction().evaluate(rule), 0);
    }

    /**
     * Tests the functionality of the evaluate-method, when dividing by zero.
     */
    @Test
    public final void testEvaluateDivisionByZero() {
        ItemSet<NamedItem> body = new ItemSet<>();
        body.add(new NamedItem("a"));
        body.setSupport(0.5);
        ItemSet<NamedItem> head = new ItemSet<>();
        head.add(new NamedItem("b"));
        AssociationRule<NamedItem> rule = new AssociationRule<>(body, head, 0.5);
        assertEquals(0, new Conviction().evaluate(rule), 0);
    }

    /**
     * Ensures, that an {@link IllegalArgumentException} is thrown by the evaluate-method, when
     * passing null as a parameter.
     */
    @Test(expected = IllegalArgumentException.class)
    public final void testEvaluateThrowsException() {
        new Conviction().evaluate(null);
    }

    /**
     * Tests the functionality of the minValue-method.
     */
    @Test
    public final void testMinValue() {
        assertEquals(0, new Conviction().minValue(), 0);
    }

    /**
     * Tests the functionality of the maxValue-method.
     */
    @Test
    public final void testMaxValue() {
        assertEquals(Double.MAX_VALUE, new Conviction().maxValue(), 0);
    }

}

File: src/test/java/de/mrapp/apriori/metrics/ConvictionTest.kt
/*
 * Copyright 2017 Michael Rapp
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package de.mrapp.apriori.metrics

import de.mrapp.apriori.AssociationRule
import de.mrapp.apriori.ItemSet
import de.mrapp.apriori.NamedItem
import kotlin.test.Test
import kotlin.test.assertEquals

/**
 * Tests the functionality of the class [Conviction].
 *
 * @author Michael Rapp
 */
class ConvictionTest {

    @Test
    fun testEvaluate() {
        val bodySupport = 0.7
        val body = ItemSet<NamedItem>()
        body.add(NamedItem("a"))
        body.support = bodySupport
        val headSupport = 0.8
        val head = ItemSet<NamedItem>()
        head.add(NamedItem("b"))
        head.support = headSupport
        val support = 0.5
        val rule = AssociationRule(body, head, support)
        val confidence = Confidence().evaluate(rule)
        assertEquals((1 - headSupport) / (1 - confidence), Conviction().evaluate(rule))
    }

    @Test
    fun testEvaluateDivisionByZero() {
        val body = ItemSet<NamedItem>()
        body.add(NamedItem("a"))
        body.support = 0.5
        val head = ItemSet<NamedItem>()
        head.add(NamedItem("b"))
        val rule = AssociationRule(body, head, 0.5)
        assertEquals(0.0, Conviction().evaluate(rule))
    }

    @Test
    fun testMinValue() {
        assertEquals(0.0, Conviction().minValue())
    }

    @Test
    fun testMaxValue() {
        assertEquals(Double.MAX_VALUE, Conviction().maxValue())
    }

}